src/theory/ff/parse.cpp

```diff
@@ -0,0 +1,433 @@
+ * fails.
+ */
+
+#include "theory/ff/parse.h"
+
+// external includes
+
+// std includes
+#include <numeric>
+#include <optional>
+#include <unordered_map>
+#include <unordered_set>
+
+// internal includes
+#include "theory/ff/util.h"
+#include "theory/theory.h"
+
+namespace cvc5::internal {
+namespace theory {
+namespace ff {
+namespace parse {
+
+// fwd declarations for Spectrum
+namespace {
+
+/**
+ * Characterization of a univariate, degree <= 2 polynomial.
+ * The members uniquely determine such a polynomial, up to its leading
+ * coefficient.
+ *
+ * This is a helper class for parsing terms that constrain a variable to be: 0,
+ * 1, or both.
+ */
+struct Spectrum
+{
+  /** the variable; ignore if degree is 0 */
+  Node var;
+  /** the degree in {0, 1, 2} */
+  uint8_t degree;
+  /** value at 0 */
+  FiniteFieldValue val0;
+  /** value at 1 */
+  FiniteFieldValue val1;
+};
+
+using SpectrumOpt = std::optional<Spectrum>;
+
+/**
+ * Attempt to compute a Spectrum for the polynomial defined by t.
+ *
+ * @param t a field term
+ * @param depth how deep to search in term before giving up
+ * @return none if t is too deep or mulitvariate or has degreee > 2; otherwise,
+ * a Spectrum.
+ */
+SpectrumOpt spectrum(const Node& t, uint8_t depth = 5);
+
+}  // namespace
+
+bool zeroConstraint(const Node& fact)
+{
+  SpectrumOpt r = spectrum(fact);
+  return r.has_value() && r->degree == 1 && r->val0.getValue() == 0;
+}
+
+bool oneConstraint(const Node& fact)
+{
+  SpectrumOpt r = spectrum(fact);
+  return r.has_value() && r->degree == 1 && r->val1.getValue() == 0;
+}
+
+std::optional<Node> bitConstraint(const Node& fact)
+{
+  SpectrumOpt r = spectrum(fact);
+  if (r.has_value() && r->degree == 2 && r->val0.getValue() == 0
+      && r->val1.getValue() == 0)
+  {
+    return {r->var};
+  }
+  return {};
+}
+
+namespace {
+
+/**
+ * Given spectra for f and g, compute an (optional) spectrum for f @ g, where @
+ * is point-wise operation
+ *
+ * @param a spectrum for f
+ * @param b spectrum for g
+ * @param fOp the binary point-wise operator @
+ * @param dOp binary operator on uint8_t for computing the degree of f @ g
+ *
+ * */
+template <typename DegreeOp, typename FieldOp>
+SpectrumOpt spectrumOp(SpectrumOpt&& a,
+                       SpectrumOpt&& b,
+                       DegreeOp dOp,
+                       FieldOp fOp)
+{
+  if (!(a.has_value() && b.has_value()))
+  {
+    // failed child
+    return {};
+  }
+  if (a->degree && b->degree && a->var != b->var)
+  {
+    // multivariate
+    return {};
+  }
+  uint8_t degree = dOp(a->degree, b->degree);
+  if (degree > 2)
+  {
+    // high-degree
+    return {};
+  }
+  FiniteFieldValue val0 = fOp(a->val0, b->val0);
+  FiniteFieldValue val1 = fOp(a->val1, b->val1);
+  Node&& var = a->degree ? std::move(a->var) : std::move(b->var);
+  return {{var, degree, val0, val1}};
+};
+
+/** Subtract spectra */
+SpectrumOpt spectrumSub(SpectrumOpt&& a, SpectrumOpt&& b)
```

Comment:
They aren't going to be used elsewhere. I made them named functions (rather than inline uses of lambdas) for readability, but if you don't think that helps, I'm happy to inline them.

Reply:
I think it's simpler to inline. As is there are two levels of indirection.